File: books/category4programmers/1.1.main.fs
let mem f = 
    let mutex = new System.Threading.Mutex()
    let mutable value = Option.None
    fun () -> 
        match value with
        | Some(v) -> v
        | None -> 
                mutex.WaitOne() |> ignore
                try
                    let v = f ()
                    value = Some v |> ignore
                    v
                finally
                    mutex.ReleaseMutex()

let Main =
    let f (x:double) = int x
    let g x = x.ToString()
    let compose f g = f >> g;
    let composed1 = f >> g
    let composed2 = g << f
    let composed3 x = x |> f |> g
    let composed4 x = f <| x |> g
    printfn "%s" <| composed1 11.0
    printfn "%s" <| composed2 22.0
    printfn "%s" <| composed3 33.0
    printfn "%s" <| composed4 44.0
    let id x = x
    printfn "%d" <| id 10
    let f1 = compose f id
    let f2 = compose id f
    printfn "%b" <| (=) (f1 11.0) (f2 11.0)
    let slowf = fun () -> 56
    let memslowf = mem slowf
    printfn "%d" <| memslowf ()

File: books/category4programmers/1.1.main.hs
import Data.Maybe

id :: a -> a
id x = x
f :: Float -> Int
f x = round x
g :: Int -> String
g x = show x
composed :: Float -> String
composed = g . f
compose :: (a->b) -> (b->c) -> (a->c)
compose f g = g . f
-- not working
mem :: (()->a) -> () -> a
mem f = 
    \() -> 
        do  
            value <- get
            case value of 
                Just v -> return v
                Nothing ->
                    put $ Some value
                    return value
                    where value = f ()
main = do
    print $ composed (Main.id 11.0)
    let f1 = compose Main.id f
    let f2 = compose f Main.id
    print $ (f1 1) == (f2 1)
    let slowf = 65
    let memslowf = mem slowf
    print $ memslowf ()